### PR TITLE
feat: unit tests for cosmos staking validators

### PR DESCRIPTION
**Pull request by chxash**
## Description

Implementation of missing unit tests for https://github.com/shapeshift/web/pull/1375

## Notice

- [X] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [X] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

https://github.com/shapeshift/web/issues/1267

### CODE REVIEW

File: src/state/slices/stakingDataSlice/stakingDataSlice.test.ts [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+      describe('validators and staking data are loaded', () => {
+        beforeEach(() => {
+          store.dispatch(
+            stakingData.actions.upsertValidatorData({
+              validators: mockValidatorData
+            })
+          )
+        })
+
+        it('returns null when validator address is unknown', async () => {
```
Comment (gomesalexandre) [inline src/state/slices/stakingDataSlice/stakingDataSlice.test.ts:318]: What does unknown mean from a code perspective? Validator not present in the slice? From Cosmos/unchained/chain-adapters perspective, it is not unknown
image

Comment (gomesalexandre) [reply src/state/slices/stakingDataSlice/stakingDataSlice.test.ts:318]:
```suggestion
        it('returns null when validator data is not present in state', async () => {
```

File: src/state/slices/stakingDataSlice/stakingDataSlice.test.ts [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+
+          const selected = selectSingleValidator(
+            store.getState(),
+            cosmosAccountSpecifier,
+            'cosmosvaloper1xxjvzwjpsf6ktuffkcpx29ut9qfrn0my8xdtqd'
+          )
+          expect(selected).toBeNull()
+        })
+
+        it('returns validator info when validator address is known', async () => {
```
Comment (gomesalexandre) [inline src/state/slices/stakingDataSlice/stakingDataSlice.test.ts:334]: Ditto naming

File: src/state/slices/stakingDataSlice/stakingDataSlice.test.ts [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+          )
+          store.dispatch(
+            stakingData.actions.upsertStakingData({
+              accountSpecifier: cosmosAccountSpecifier,
+              stakingData: mockStakingData
+            })
+          )
+        })
+
+        it('returns empty string when there is no reward', async () => {
```
Comment (gomesalexandre) [inline src/state/slices/stakingDataSlice/stakingDataSlice.test.ts:383]: No rewards != no rewards data, you could have rewards on-chain but for xyz reason related to our business logic the store isn't populated with it
```suggestion
        it('returns empty string when there is no rewards data', async () => {
```

File: src/state/slices/stakingDataSlice/stakingDataSlice.test.ts [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+            SHAPESHIFT_VALIDATOR_ADDRESS,
+            cosmosAssetId
+          )
+          expect(selected).toEqual('3.831752143667562385')
+        })
+      })
+    })
+
+    describe('selectTotalBondingsBalanceByAssetId', () => {
+      it('returns 0 when state is empty', async () => {
```
Comment (gomesalexandre) [inline src/state/slices/stakingDataSlice/stakingDataSlice.test.ts:414]: Making it clear we return numbers as strings
```suggestion
      it('returns '0' when state is empty', async () => {
```

File: src/state/slices/stakingDataSlice/stakingDataSlice.test.ts [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+            SHAPESHIFT_VALIDATOR_ADDRESS,
+            cosmosAssetId
+          )
+          expect(selected).toMatchSnapshot()
+        })
+      })
+    })
+
+    describe('selectNonloadedValidators', () => {
+      it('returns empty array when state is empty', async () => {
```
Comment (gomesalexandre) [inline src/state/slices/stakingDataSlice/stakingDataSlice.test.ts:240]:
```suggestion
      it('returns empty array on initial state', async () => {
```
I'd go with `initial state` terminology instead, empty is a high-level human concept, but from a structure perspective, the initial state is not an empty object

File: src/state/slices/stakingDataSlice/stakingDataSlice.test.ts [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+            apr: '0.1496681491',
+            commission: '0.100000000000000000',
+            moniker: 'ShapeShift DAO'
+          })
+        })
+      })
+    })
+
+    describe('selectRewardsAmountByAssetId', () => {
+      it('returns empty string when state is empty', async () => {
```
Comment (gomesalexandre) [inline src/state/slices/stakingDataSlice/stakingDataSlice.test.ts:358]:
```suggestion
      it('returns empty string on initial state', async () => {
```

**Review by gomesalexandre — APPROVED** [review]
LGTM!
A few non-blocking comments as suggested changes mainly regarding naming of tests, feel free to click "commit suggested changes", go with another commit that implements them in all additional places, or leave it as is.  At your own discretion

**Comment by chxash** [comment]
> LGTM! A few non-blocking comments as suggested changes mainly regarding naming of tests, feel free to click "commit suggested changes", go with another commit that implements them in all additional places, or leave it as is. At your own discretion

Fixed all comments !

**Review by 0xApotheosis — APPROVED** [review]
Nice, comprehensive tests!

**Pull request merged**